### PR TITLE
Initial version to support OWIN, .NET, .NET Core

### PR DESCRIPTION
**Pull request by jmprieur**
Comments to add - not ready to review
Needs a few updates to unit tests

Not ready to review

There are a lot of files changed, but this is mainly because of the new OWIN sample

### CODE REVIEW

File: tests/daemon-app/Daemon-app/Program.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+namespace daemon_console
+{
+    /// <summary>
+    /// This sample shows how to query the Microsoft Graph from a daemon application
+    /// which uses application permissions.
+    /// For more information see https://aka.ms/msal-net-client-credentials
+    /// </summary>
+    class Program
+    {
+        static void Main(string[] args)
```
Comment (bgavrilMS) [inline tests/daemon-app/Daemon-app/Program.cs:19]: `static async Task Main(string[] args)` will  work just as well.

File: src/Microsoft.Identity.Web/ITokenAcquirer.cs [hunk 2]
```diff
@@ -21,18 +20,18 @@ public interface ITokenAcquirer
...
         /// user (in a web app), or the user for which the token was received (in a web API)
         /// cases where a given account is a guest in other tenants, and you want to acquire tokens for a specific tenant, like where the user is a guest in.</param>
-        /// <param name="tokenAcquisitionOptions">Options passed-in to create the token acquisition object which calls into MSAL .NET.</param>
         /// <returns>An <see cref="AuthenticationResult"/> to call on behalf of the user, the downstream API characterized by its scopes.</returns>
         Task<AuthenticationResult> GetAuthenticationResultForUserAsync(
             IEnumerable<string> scopes,
+            TokenAcquisitionOptions? tokenAcquisitionOptions = null,
             string? authenticationScheme = null,
-            ClaimsPrincipal? user = null,
-            TokenAcquisitionOptions? tokenAcquisitionOptions = null);
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web/ITokenAcquirer.cs]: I assume this interface has not been published before, as reordering params is a breaking change.

File: src/Microsoft.Identity.Web/Hosts/PlainDotNet/PlainDotNetTokenAcquisitionHost.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+﻿using System.Net;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Client;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Identity.Web.Hosts
+{
+    internal class PlainDotNetTokenAcquisitionHost : ITokenAcquisitionHost
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web/Hosts/PlainDotNet/PlainDotNetTokenAcquisitionHost.cs:10]: Maybe `DefaultTokenAcquisitionHost` or `NullTokenAcquisitionHost`

File: src/Microsoft.Identity.Web/Hosts/PlainDotNet/PlainDotNetTokenAcquisitionHost.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            IOptionsMonitor<MicrosoftIdentityOptions> optionsMonitor, 
+            IOptionsMonitor<MergedOptions> mergedOptionsMonitor,
+            IOptionsMonitor<ConfidentialClientApplicationOptions> ccaOptionsMonitor)
+        {
+            _optionsMonitor = optionsMonitor;
+            _mergedOptionsMonitor = mergedOptionsMonitor;
+            _ccaOptionsMonitor = ccaOptionsMonitor;
+        }
+
+        Task<ClaimsPrincipal?> ITokenAcquisitionHost.GetAuthenticatedUserAsync(ClaimsPrincipal? user)
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web/Hosts/PlainDotNet/PlainDotNetTokenAcquisitionHost.cs:27]: What is the reason for using explicit interface implementation? Is it to be able to keep this class internal? Needs a comment ..

File: changelog.md [hunk 5]
```diff
@@ -23,6 +23,21 @@
 ### Bug Fixes:
 **Microsoft.Identity.Web now returns `TokenValidatedContext.Fail` instead of throwing `UnauthorizedAccessException` in case of missing roles or scopes**, which enables a better developer experience. See issue [#1716](https://github.com/AzureAD/microsoft-identity-web/issues/1716) for details.
 
+1.25.0
+==========
```
Comment (jennyf19) [inline changelog.md:27]: This is strange?

File: build/template-install-dotnet-core.yaml [hunk 6]
```diff
@@ -12,6 +12,10 @@ steps:
   inputs:
     version: 6.0.x   
 
+- task: UseDotNet@2
+  displayName: 'Use .Net Core SDK 6'
+  inputs:
+    version: 6.0.x
```
Comment (jennyf19) [inline build/template-install-dotnet-core.yaml:18]: I think we already have this above

File: src/Microsoft.Identity.Web/WebAppExtensions/MicrosoftIdentityAppCallingWebApiAuthenticationBuilderExt.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+using Microsoft.Identity.Web.TokenCacheProviders;
+using Microsoft.Identity.Web.TokenCacheProviders.Distributed;
+using Microsoft.Identity.Web.TokenCacheProviders.InMemory;
+using Microsoft.Identity.Web.TokenCacheProviders.Session;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Authentication builder returned by the EnableTokenAcquisitionToCallDownstreamApi methods
+    /// enabling you to decide token cache implementations.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/WebAppExtensions/MicrosoftIdentityAppCallingWebApiAuthenticationBuilderExt.cs:22]: >token cache implementations [](http://example.com/codeflow?start=31&length=27)

specify this is only for session cache

File: src/Microsoft.Identity.Web/WebAppExtensions/MicrosoftIdentityWebAppAuthenticationBuilder.cs [hunk 8]
```diff
@@ -172,7 +172,7 @@ public class MicrosoftIdentityWebAppAuthenticationBuilder : MicrosoftIdentityBas
                        {
                            // Remove the account from MSAL.NET token cache
                            var tokenAcquisition = context!.HttpContext.RequestServices.GetRequiredService<ITokenAcquisitionInternal>();
-                           await tokenAcquisition.RemoveAccountAsync(context, openIdConnectScheme).ConfigureAwait(false);
+                           await tokenAcquisition.RemoveAccountAsync(context!.HttpContext.User, openIdConnectScheme).ConfigureAwait(false);
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/WebAppExtensions/MicrosoftIdentityWebAppAuthenticationBuilder.cs:175]: >context! [](http://example.com/codeflow?start=69&length=8)

do we need to lock on the context here?

File: src/Microsoft.Identity.Web.MicrosoftGraph/Microsoft.Identity.Web.MicrosoftGraph.csproj [hunk 9]
```diff
@@ -20,7 +20,11 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\Microsoft.Identity.Web\Microsoft.Identity.Web.csproj" />
+    <ProjectReference Include="..\Microsoft.Identity.Web.TokenAcquisition\Microsoft.Identity.Web.TokenAcquisition.csproj" />
   </ItemGroup>
 
+  <PropertyGroup>
+    <TargetFrameworks>netcoreapp3.1; net5.0; net462; net472</TargetFrameworks>
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.MicrosoftGraph/Microsoft.Identity.Web.MicrosoftGraph.csproj:27]: >net5.0 [](http://example.com/codeflow?start=37&length=6)

this is related from line 15, and should we support net6 instead?

File: src/Microsoft.Identity.Web.MicrosoftGraph/MicrosoftGraphExtensions.cs [hunk 10]
```diff
@@ -11,6 +11,7 @@
 
 namespace Microsoft.Identity.Web
 {
+#if !NET472 && !NET462
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.MicrosoftGraph/MicrosoftGraphExtensions.cs:14]: >#if !N [](http://example.com/codeflow?start=0&length=6)

let's discuss

File: src/Microsoft.Identity.Web.OWIN/AppBuilderExtension.cs [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Extension methods on an ASP.NET application to add a web app or web API
+    /// </summary>
+    public static class AppBuilderExtension
+    {
+        static internal TokenAcquirerFactory? TokenAcquirerFactory { get; set; }
+        /// <summary>
+        /// Configuration
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.OWIN/AppBuilderExtension.cs:27]: nit: line

**Comment by jennyf19** [comment]
> using System.Web;

license info

---
Refers to: src/Microsoft.Identity.Web.OWIN/HttpContextExtensions.cs:1 in 29bb54a. [](commit_id = 29bb54abc6fc26b481fc7ba694133e47d05f10f8, deletion_comment = False)

**Comment by jennyf19** [comment]
> using Microsoft.IdentityModel.Protocols;

license info

---
Refers to: src/Microsoft.Identity.Web.OWIN/OpenIdConnectCachingSecurityTokenProvider.cs:1 in 29bb54a. [](commit_id = 29bb54abc6fc26b481fc7ba694133e47d05f10f8, deletion_comment = False)

**Comment by jennyf19** [comment]
> using System.Net;

license info

---
Refers to: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:1 in 29bb54a. [](commit_id = 29bb54abc6fc26b481fc7ba694133e47d05f10f8, deletion_comment = False)

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.Identity.Client;
+using Microsoft.IdentityModel.JsonWebTokens;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Identity.Web.Hosts
+{
+    internal class OwinTokenAcquisitionHost : ITokenAcquisitionHost
+    {
+
+        IOptionsMonitor<MicrosoftIdentityOptions> _microsoftIdentityOptionsMonitor;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:15]: nit: remove line

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        IOptionsMonitor<MergedOptions> _mergedOptionsMonitor;
+        IOptionsMonitor<ConfidentialClientApplicationOptions> _ccaOptionsMonitor;
+
+        public OwinTokenAcquisitionHost(
+            IOptionsMonitor<MicrosoftIdentityOptions> optionsMonitor,
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:20]: >optionsMonitor [](http://example.com/codeflow?start=54&length=14)

better to call these MicrosoftIdentityOption?

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        public Task<ClaimsPrincipal?> GetAuthenticatedUserAsync(ClaimsPrincipal? user)
+        {
+            return Task.FromResult<ClaimsPrincipal?>(null);
+        }
+
+        public string? GetCurrentRedirectUri(MergedOptions mergedOptions)
+        {
+            return null;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:36]: >return null [](http://example.com/codeflow?start=12&length=11)

this should probably do something?

**Comment by jennyf19** [comment]
> using System.Net;

license info

---
Refers to: src/Microsoft.Identity.Web.TokenAcquisition/ITokenAcquisitionHost.cs:1 in 29bb54a. [](commit_id = 29bb54abc6fc26b481fc7ba694133e47d05f10f8, deletion_comment = False)

**Comment by jennyf19** [comment]
> using System.Net;

licnese info

---
Refers to: src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs:1 in 29bb54a. [](commit_id = 29bb54abc6fc26b481fc7ba694133e47d05f10f8, deletion_comment = False)

File: src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
+﻿using System.Net;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Client;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Microsoft.Identity.Web.Hosts
+{
+    internal class PlainDotNetTokenAcquisitionHost : ITokenAcquisitionHost
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs:10]: let's think of a better name

**Comment by jennyf19** [comment]
> using System;

license info

---
Refers to: src/Microsoft.Identity.Web.TokenAcquisition/AspNetCore/TokenAcquisition-AspnetCore.cs:1 in 29bb54a. [](commit_id = 29bb54abc6fc26b481fc7ba694133e47d05f10f8, deletion_comment = False)

**Comment by jennyf19** [comment]
I'm only partway through. I will the branch and continue tomorrow

File: changelog.md (hunk 5 above)
Comment (jennyf19) [reply changelog.md:27]: fixed.

File: src/Microsoft.Identity.Web/WebAppExtensions/MicrosoftIdentityAppCallingWebApiAuthenticationBuilderExt.cs (hunk 7 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web/WebAppExtensions/MicrosoftIdentityAppCallingWebApiAuthenticationBuilderExt.cs:22]: done.

File: src/Microsoft.Identity.Web.MicrosoftGraph/Microsoft.Identity.Web.MicrosoftGraph.csproj (hunk 9 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.MicrosoftGraph/Microsoft.Identity.Web.MicrosoftGraph.csproj:27]: done. but not .net 6.

File: src/Microsoft.Identity.Web.OWIN/AppBuilderExtension.cs (hunk 11 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.OWIN/AppBuilderExtension.cs:27]: fixed.

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs (hunk 12 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:15]: fixed.

File: build/template-install-dotnet-core.yaml (hunk 6 above)
Comment (jennyf19) [reply build/template-install-dotnet-core.yaml:18]: fixed.

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs (hunk 13 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:22]: yes!

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs (hunk 14 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:38]: yes, I agree. The web app part wasn't implemented yet.

File: src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs (hunk 15 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs:13]: DefaultTokenAcquisitionHost seems good indeed.

File: src/Microsoft.Identity.Web.MicrosoftGraph/MicrosoftGraphExtensions.cs (hunk 10 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.MicrosoftGraph/MicrosoftGraphExtensions.cs:14]: needs to stay for now.

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs (hunk 13 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:22]: updated.

File: src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs (hunk 14 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.OWIN/OwinTokenAcquisitionHost.cs:38]: fixed.

File: src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs (hunk 15 above)
Comment (jennyf19) [reply src/Microsoft.Identity.Web.TokenAcquisition/PlainDotNetTokenAcquisitionHost.cs:13]: renamed.

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Pull request merged**